src/OrleansAzureUtils/MultiClusterNetwork/AzureTableBasedGossipChannel.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Runtime.Configuration;
+using Orleans.MultiCluster;
+
+namespace Orleans.Runtime.MultiClusterNetwork
+{
+    /// <summary>
+    /// An implementation of a gossip channel based on a standard Orleans Azure table.
+    /// Multiple gossip networks can use the same table, and are separated by partition key = GlobalServiceId
+    /// </summary>
+    internal class AzureTableBasedGossipChannel : IGossipChannel
+    {
+        private TraceLogger logger;
+        private GossipTableInstanceManager tableManager;
+        private static int sequenceNumber;
+
+        public string Name { get; private set; }
+
+        public async Task Initialize(GlobalConfiguration globalconfig, string connectionstring)
+        {
+            Name = "AzureTableBasedGossipChannel-" + ++sequenceNumber;
+            logger = TraceLogger.GetLogger(Name, TraceLogger.LoggerType.Runtime);
+
+            logger.Info("Initializing Gossip Channel for ServiceId={0} using connection: {1}, SeverityLevel={2}",
+                globalconfig.GlobalServiceId, ConfigUtilities.RedactConnectionStringInfo(connectionstring), logger.SeverityLevel);
+
+            tableManager = await GossipTableInstanceManager.GetManager(globalconfig.GlobalServiceId, connectionstring, logger);
+        }
+
+        // used by unit tests
+        public Task DeleteAllEntries()
+        {
+            logger.Info("DeleteAllEntries");
+            return tableManager.DeleteTableEntries();
+        }
+
+      
+        private static void UpdateDictionaryRightValue<TKey, TLeft, TRight>(Dictionary<TKey, KeyValuePair<TLeft, TRight>> dict, TKey key, TRight rightValue)
+        {
+            if (dict.ContainsKey(key))
+                dict[key] = new KeyValuePair<TLeft, TRight>(dict[key].Key, rightValue);
+            else
+                dict.Add(key, new KeyValuePair<TLeft, TRight>(default(TLeft), rightValue));
+        }
+
+        // IGossipChannel
+        public async Task Push(MultiClusterData data)
+        {
+            logger.Verbose("-Push data:{0}", data);
+
+            var retrievalTasks = new List<Task<GossipTableEntry>>();
+            if (data.Configuration != null)
+            {
+                retrievalTasks.Add(tableManager.ReadConfigurationEntryAsync());
+            }
+
+            foreach (var gateway in data.Gateways.Values)
+            {
+                retrievalTasks.Add(tableManager.ReadGatewayEntryAsync(gateway));
+            }
+
+            await Task.WhenAll(retrievalTasks);
```

Comment:
What's confusing to me here is that tasks returned by two different calls,  `ReadConfigurationEntryAsync` and `ReadGatewayEntryAsync` get put into a single list, and their results are then process together. I think it's better to be explicit and pass the value returned by `ReadConfigurationEntryAsync` to `DiffAndWriteBack` as a separate argument instead of mixing it with different items returned by `ReadGatewayEntryAsync` calls.